nx/lib/nx.ex

```diff
@@ -7047,6 +7047,185 @@ defmodule Nx do
+    do: cumulative_op(tensor, opts, :cumulative_sum, :window_sum)
+
+  @doc """
+  Returns the cumulative product of elements along an axis.
+
+  ## Options
+
+    * `:axis` - the axis to multiply elements along. Defaults to `0`
+
+  ## Examples
+
+      iex> Nx.cumulative_product(Nx.tensor([1, 2, 3, 4]))
+      #Nx.Tensor<
+        s64[4]
+        [1, 2, 6, 24]
+      >
+
+      iex> Nx.cumulative_product(Nx.iota({3, 3}), axis: 0)
+      #Nx.Tensor<
+        s64[3][3]
+        [
+          [0, 1, 2],
+          [0, 4, 10],
+          [0, 28, 80]
+        ]
+      >
+
+      iex> Nx.cumulative_product(Nx.iota({3, 3}), axis: 1)
+      #Nx.Tensor<
+        s64[3][3]
+        [
+          [0, 0, 0],
+          [3, 12, 60],
+          [6, 42, 336]
+        ]
+      >
+  """
+  @doc type: :cumulative
+  def cumulative_product(tensor, opts \\ []),
+    do: cumulative_op(tensor, opts, :cumulative_product, :window_product)
+
+  @doc """
+  Returns the cumulative minimum of elements along an axis.
+
+  ## Options
+
+    * `:axis` - the axis to compare elements along. Defaults to `0`
+
+  ## Examples
+
+      iex> Nx.cumulative_min(Nx.tensor([3, 4, 2, 1]))
+      #Nx.Tensor<
+        s64[4]
+        [3, 3, 2, 1]
+      >
+
+      iex> Nx.cumulative_min(Nx.tensor([[2, 3, 1], [1, 3, 2], [2, 1, 3]]), axis: 0)
+      #Nx.Tensor<
+        s64[3][3]
+        [
+          [2, 3, 1],
+          [1, 3, 1],
+          [1, 1, 1]
+        ]
+      >
+
+      iex> Nx.cumulative_min(Nx.tensor([[2, 3, 1], [1, 3, 2], [2, 1, 3]]), axis: 1)
+      #Nx.Tensor<
+        s64[3][3]
+        [
+          [2, 2, 1],
+          [1, 1, 1],
+          [2, 1, 1]
+        ]
+      >
+  """
+  @doc type: :cumulative
+  def cumulative_min(tensor, opts \\ []),
+    do: cumulative_op(tensor, opts, :cumulative_min, :window_min)
+
+  @doc """
+  Returns the cumulative maximum of elements along an axis.
+
+  ## Options
+
+    * `:axis` - the axis to compare elements along. Defaults to `0`
+
+  ## Examples
+
+      iex> Nx.cumulative_max(Nx.tensor([3, 4, 2, 1]))
+      #Nx.Tensor<
+        s64[4]
+        [3, 4, 4, 4]
+      >
+
+      iex> Nx.cumulative_max(Nx.tensor([[2, 3, 1], [1, 3, 2], [2, 1, 3]]), axis: 0)
+      #Nx.Tensor<
+        s64[3][3]
+        [
+          [2, 3, 1],
+          [2, 3, 2],
+          [2, 3, 3]
+        ]
+      >
+
+      iex> Nx.cumulative_max(Nx.tensor([[2, 3, 1], [1, 3, 2], [2, 1, 3]]), axis: 1)
+      #Nx.Tensor<
+        s64[3][3]
+        [
+          [2, 3, 3],
+          [1, 3, 3],
+          [2, 2, 3]
+        ]
+      >
+  """
+  @doc type: :cumulative
+  def cumulative_max(tensor, opts \\ []),
+    do: cumulative_op(tensor, opts, :cumulative_max, :window_max)
+
+  defp cumulative_op(tensor, opts, op, window_op) do
+    opts = keyword!(opts, axis: 0)
+    tensor = to_tensor(tensor)
+    axis = Nx.Shape.normalize_axis(tensor.shape, opts[:axis], tensor.names)
+
+    Nx.Shared.optional(op, [tensor, axis], tensor, fn tensor, axis ->
```

Comment:
@josevalim I defined the callbacks as

```elixir
@callback cumulative_sum(out :: tensor, t :: tensor, axis) :: tensor
```

however, we can't pass axis this way, because `EXLA.Backend` does JIT and converts `axis` to tensor. Thoughts on what approach to take?

Reply:
```suggestion
    Nx.Shared.optional(op, [tensor, axis], tensor, fn tensor, _axis ->
```

This would work, because we would capture and use the axis directly, it's just a little bit weird.

Reply:
@jonatanklosko we should change EXLA.Backend to JIT only on the first argument. We are being too eager on the JITing there.

Reply:
Ahhh, we have the callback name, so even if we add more tensors, we can just differentiate based on the name, got it!

Reply:
Yeah, but even then I think all optional callbacks only need to JIT on the first arguyment!

Reply:
Hmm, what do you mean?

```elixir
@callback solve(out :: tensor, a :: tensor, b :: tensor) :: tensor
```

We want to JIT on both `a` and `b`, no?

Reply:
Updated!